File: grahamcraft2/client/player_model.py
"""Low-poly blocky avatars for local and remote players."""

from __future__ import annotations

from dataclasses import dataclass
from hashlib import sha256
from typing import Final

from ursina import Entity, Vec3, camera, color, destroy

from grahamcraft2.client.arrow_key_controller import ArrowKeyController
from grahamcraft2.client.voxels import BLOCK_PALETTE

CAMERA_BACK_OFFSET: Final = Vec3(0, 0.25, -6)


@dataclass(frozen=True, slots=True)
class AvatarColors:
    """Tint colors for a blocky character."""

    skin: color
    body: color
    limbs: color


@dataclass(frozen=True, slots=True)
class _BodyPart:
    """One cube in the blocky character rig."""

    position: Vec3
    scale: Vec3
    slot: str


_BODY_PARTS: Final = (
    _BodyPart(Vec3(0, 1.55, 0), Vec3(0.45, 0.45, 0.45), "skin"),
    _BodyPart(Vec3(0, 1.0, 0), Vec3(0.5, 0.65, 0.3), "body"),
    _BodyPart(Vec3(-0.15, 0.35, 0), Vec3(0.22, 0.7, 0.22), "limbs"),
    _BodyPart(Vec3(0.15, 0.35, 0), Vec3(0.22, 0.7, 0.22), "limbs"),
    _BodyPart(Vec3(-0.38, 1.0, 0), Vec3(0.2, 0.6, 0.2), "body"),
    _BodyPart(Vec3(0.38, 1.0, 0), Vec3(0.2, 0.6, 0.2), "body"),
)

LOCAL_AVATAR_COLORS: Final = AvatarColors(
    skin=color.rgb32(255, 213, 170),
    body=color.azure,
    limbs=color.blue,
)


def colors_for_player(player_id: str) -> AvatarColors:
    """Pick a stable palette color set from a player id."""
    digest = sha256(player_id.encode()).digest()
    body = BLOCK_PALETTE[digest[0] % len(BLOCK_PALETTE)]
    limbs = BLOCK_PALETTE[digest[1] % len(BLOCK_PALETTE)]
    skin = color.rgb32(255, 213, 170)
    return AvatarColors(skin=skin, body=body, limbs=limbs)


def _tint_for_slot(colors: AvatarColors, slot: str) -> color:
    if slot == "skin":
        return colors.skin
    if slot == "limbs":
        return colors.limbs
    return colors.body


class BlockyFigure:
    """A cube-built character rig in the world."""

    def __init__(
        self,
        position: Vec3,
        colors: AvatarColors,
        parent: Entity | None = None,
    ) -> None:
        """Create a blocky avatar at the given position."""
        self.root = Entity(parent=parent, position=position)
        self.parts: list[Entity] = []
        for part in _BODY_PARTS:
            cube = Entity(
                parent=self.root,
                model="cube",
                position=part.position,
                scale=part.scale,
                texture="white_cube",
                color=_tint_for_slot(colors, part.slot),
                collider=None,
                shader=None,
            )
            self.parts.append(cube)

    def set_position(self, x: float, y: float, z: float) -> None:
        """Move the avatar root to a server position."""
        self.root.position = Vec3(x, y, z)

    def remove(self) -> None:
        """Destroy the avatar from the scene."""
        destroy(self.root)
        self.parts.clear()


class PlayerAvatar:
    """Visible blocky character following the local player."""

    def __init__(self, controller: ArrowKeyController) -> None:
        """Build a cube rig and switch to a third-person camera."""
        self.controller = controller
        self.figure = BlockyFigure(
            position=Vec3(0, 0, 0),
            colors=LOCAL_AVATAR_COLORS,
            parent=controller,
        )
        controller.ignore_list.extend(self.figure.parts)
        controller.ignore_list.append(self.figure.root)
        camera.position = CAMERA_BACK_OFFSET

    @property
    def raycast_ignore(self) -> list[Entity]:
        """Entities to exclude from placement and ground raycasts."""
        return self.figure.parts

File: client/src/player_model.rs
//! Low-poly blocky avatars for local and remote players.

use bevy::prelude::*;
use sha2::{Digest, Sha256};

use crate::models::{block_color, RaycastIgnore, BLOCK_PALETTE};

const SKIN: Color = Color::srgb(1.0, 0.835, 0.667);

#[derive(Clone, Copy)]
pub(crate) struct AvatarColors {
    skin: Color,
    body: Color,
    limbs: Color,
}

const LOCAL_AVATAR: AvatarColors = AvatarColors {
    skin: SKIN,
    body: Color::srgb(0.0, 0.5, 1.0),
    limbs: Color::srgb(0.0, 0.0, 1.0),
};

struct BodyPart {
    position: Vec3,
    scale: Vec3,
    slot: &'static str,
}

const BODY_PARTS: [BodyPart; 6] = [
    BodyPart {
        position: Vec3::new(0.0, 1.55, 0.0),
        scale: Vec3::new(0.45, 0.45, 0.45),
        slot: "skin",
    },
    BodyPart {
        position: Vec3::new(0.0, 1.0, 0.0),
        scale: Vec3::new(0.5, 0.65, 0.3),
        slot: "body",
    },
    BodyPart {
        position: Vec3::new(-0.15, 0.35, 0.0),
        scale: Vec3::new(0.22, 0.7, 0.22),
        slot: "limbs",
    },
    BodyPart {
        position: Vec3::new(0.15, 0.35, 0.0),
        scale: Vec3::new(0.22, 0.7, 0.22),
        slot: "limbs",
    },
    BodyPart {
        position: Vec3::new(-0.38, 1.0, 0.0),
        scale: Vec3::new(0.2, 0.6, 0.2),
        slot: "body",
    },
    BodyPart {
        position: Vec3::new(0.38, 1.0, 0.0),
        scale: Vec3::new(0.2, 0.6, 0.2),
        slot: "body",
    },
];

fn colors_for_player(player_id: &str) -> AvatarColors {
    let digest = Sha256::digest(player_id.as_bytes());
    AvatarColors {
        skin: SKIN,
        body: block_color((digest[0] as i32 % BLOCK_PALETTE.len() as i32) + 1),
        limbs: block_color((digest[1] as i32 % BLOCK_PALETTE.len() as i32) + 1),
    }
}

fn tint_for_slot(colors: AvatarColors, slot: &str) -> Color {
    match slot {
        "skin" => colors.skin,
        "limbs" => colors.limbs,
        _ => colors.body,
    }
}

pub fn spawn_blocky_avatar(
    commands: &mut Commands,
    cube_mesh: Handle<Mesh>,
    materials: &mut Assets<StandardMaterial>,
    parent: Entity,
    colors: AvatarColors,
) -> Entity {
    let root = commands
        .spawn((
            Transform::default(),
            Visibility::default(),
        ))
        .id();
    commands.entity(parent).add_child(root);

    for part in BODY_PARTS {
        let material = materials.add(StandardMaterial {
            base_color: tint_for_slot(colors, part.slot),
            perceptual_roughness: 0.9,
            ..default()
        });
        commands.entity(root).with_children(|parent| {
            parent.spawn((
                RaycastIgnore,
                Mesh3d(cube_mesh.clone()),
                MeshMaterial3d(material),
                Transform::from_translation(part.position).with_scale(part.scale),
                Visibility::default(),
            ));
        });
    }

    root
}

pub fn spawn_local_avatar(
    commands: &mut Commands,
    cube_mesh: Handle<Mesh>,
    materials: &mut Assets<StandardMaterial>,
    player: Entity,
) {
    spawn_blocky_avatar(commands, cube_mesh, materials, player, LOCAL_AVATAR);
}

pub fn spawn_remote_avatar(
    commands: &mut Commands,
    cube_mesh: Handle<Mesh>,
    materials: &mut Assets<StandardMaterial>,
    player_id: &str,
    position: Vec3,
) -> Entity {
    let colors = colors_for_player(player_id);
    let root = commands
        .spawn((
            Transform::from_translation(position),
            Visibility::default(),
        ))
        .id();
    spawn_blocky_avatar(commands, cube_mesh, materials, root, colors);
    root
}
